test/middleware.cjs

```diff
@@ -281,6 +281,33 @@ describe('middleware', () => {
       }
     });
 
+    // Addresses: https://github.com/yargs/yargs/issues/2124
+    // This test will fail if the result of async middleware is not treated like a promise
+    it('treats result of async middleware as promise', done => {
+      const input = 'cmd1 -f Hello -b world';
+      yargs(input)
+        .command(
+          'cmd1',
+          'cmd1 desc',
+          yargs =>
+            yargs
```

Comment:
I had to read this a few times, I'm wondering if it would be worth simplifying to exactly the minimal case outlined in the regression:

```
Yargs
  .strict(true)
  .middleware(async () => {}, true)
  .command("test", false, () => {}, () => {})
  .parseAsync(['test']);
```

I don't have a strong opinion though, if your goal was to exercise parts of the codebase that weren't covered in the above example.